File: exercises/practice/diffie-hellman/.meta/src/reference/java/DiffieHellman.java
import java.math.BigInteger;
import java.util.Random;

class DiffieHellman {

    private Random random = new Random();

    BigInteger privateKey(BigInteger primeP) {
        BigInteger privateKey;
        do {
            privateKey = new BigInteger(primeP.bitLength(), random);
        } while (privateKey.compareTo(BigInteger.ZERO) == 0 || privateKey.compareTo(primeP) >= 0);
        return privateKey;
    }

    BigInteger publicKey(BigInteger primeP, BigInteger primeG, BigInteger privateKey) {
        return primeG.modPow(privateKey, primeP);
    }

    BigInteger secret(BigInteger primeP, BigInteger publicKey, BigInteger privateKey) {
        return publicKey.modPow(privateKey, primeP);
    }

}

File: go/diffie-hellman/diffie_hellman.go
package diffiehellman

import (
	"crypto/rand"
	"math/big"
)

var TWO = big.NewInt(2)

func PrivateKey(primeP *big.Int) *big.Int {
	max := big.NewInt(0).Sub(primeP, TWO)
	random, err := rand.Int(rand.Reader, max)
	if err != nil {
		panic(err)
	}
	return random.Add(TWO, random)
}

func PublicKey(private, primeP *big.Int, primeG int64) *big.Int {
	return big.NewInt(0).Exp(big.NewInt(primeG), private, primeP)
}

func NewPair(primeP *big.Int, primeG int64) (*big.Int, *big.Int) {
	private := PrivateKey(primeP)
	return private, PublicKey(private, primeP, primeG)
}

func SecretKey(private1, public2, primeP *big.Int) *big.Int {
	return big.NewInt(0).Exp(public2, private1, primeP)
}
